### PR TITLE
fix gateway bind with same port and different protocol

### PR DESCRIPTION
**Pull request by shonecyx**
fix https://github.com/istio/istio/issues/43688

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
can you elaborate how you do resolve the conflict

**Comment by shonecyx** [comment]
> can you elaborate how you do resolve the conflict

@hzxuzhonghu This is similar issue like this one: https://github.com/istio/istio/pull/40275
The idea is that, if there is bind in gateway and bind to different hosts, same port but different protocol, the listeners should be generate separately without conflict.
For example for below 3 gateways, there should be 3 listeners generated successfully: `0.0.0.0_443` for GRPC, `1.0.0.1_443` for HTTPS and `1.0.0.2_443` for GRPC.
```
gateway1, bind:"",port:443,protocol:GRPC
gateway2, bind:"1.0.0.1",port:443,protocol:HTTPS
gateway3, bind:"1.0.0.2",port:443,protocol:GRPC
```

File: releasenotes/notes/43688.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: traffic-management
+issue:
+  - https://github.com/istio/istio/issues/43688
+releaseNotes:
+  - |
+    **Fixed** an issue that when there is different Bind specified in the Gateways with same port and different protocol, listeners are not generated correctly.
```
Comment (ericvn) [inline releasenotes/notes/43688.yaml:8]:
```suggestion
    **Fixed** an issue that when there are different Binds specified in the Gateways with the same port and different protocols, listeners are not generated correctly.
```

File: pilot/pkg/model/gateway.go [hunk 2]
```diff
@@ -268,9 +269,18 @@ func MergeGateways(gateways []gatewayWithInstances, proxy *Proxy, ps *PushContex
 							}
 							serversByRouteName[routeName] = []*networking.Server{s}
 						}
-
+						if bindsMap, ok := nonPlainTextGatewayPortsBindMap[resolvedPort]; ok {
+							bindsMap[serverPort.Bind] = "true"
```
Comment (howardjohn) [inline pilot/pkg/model/gateway.go:273]: if the map is only going to container "true" it should just be a set. `map[uint32]sets.String`

Comment (shonecyx) [reply pilot/pkg/model/gateway.go]: Updated to map[uint32]sets.String, thanks!

File: releasenotes/notes/43688.yaml (hunk 1 above)
Comment (shonecyx) [reply releasenotes/notes/43688.yaml]: Updated, thanks.

**Review by ramaraochavali — APPROVED** [review]

**Review by ericvn — APPROVED** [review]
Release notes and test looks good.

**Pull request merged**